python/lsst/ts/m2com/mock/mock_control_closed_loop.py

```diff
@@ -475,6 +475,286 @@ def select_axial_hardpoints(location_axial_actuator, specific_axial_hardpoint):
+        displacements.
+
+        Notes
+        -----
+        Translate the calculation from the HardPointToRigidBody.vi in ts_mtm2.
+
+        Parameters
+        ----------
+        location_axial_actuator : `list [list]`
+            Location of the axial actuators: (x, y) in meter. This should be a
+            72 x 2 matrix.
+        location_tangent_link : `list`
+            Location of the tangent links in degree. This should be a 1 x 6
+            array.
+        radius : `float`
+            Radius of the cell in meter.
+        hardpoints : `list`
+            Six hardpoints. The order is from low to high.
+        disp_hardpoint_current : `list`
+            Six current hardpoint displacements. The units are meter and
+            radian.
+        disp_hardpoint_home : `list`
+            Six hardpoint displacements at the home position. The units are
+            meter and radian.
+
+        Returns
+        -------
+        x : `float`
+            X position in meter.
+        y : `float`
+            Y position in meter.
+        z : `float`
+            Z position in meter.
+        rx : `float`
+            X rotator in radian.
+        ry : `float`
+            Y rotator in radian.
+        rz : `float`
+            Z rotator in radian.
+        """
+
+        # Delta of the axial hardpoint displacements
+        num_hardpoints_axial = 3
+        disp_hardpoint_axial = (
+            np.array(disp_hardpoint_current)[:num_hardpoints_axial]
+            - np.array(disp_hardpoint_home)[:num_hardpoints_axial]
+        )
+
+        # Location of the axial hardpoints: (x_hp, y_hp, 1)
+        location_axial_actuator = np.array(location_axial_actuator)
+        location_axial_hardpoint = np.append(
+            location_axial_actuator[hardpoints[:num_hardpoints_axial], :],
+            np.ones((num_hardpoints_axial, 1)),
+            axis=1,
+        )
+
+        # The displacement of axial actuators decides the (z, rx, ry) of rigid
+        # body.
+        dof_axial = np.linalg.inv(location_axial_hardpoint).dot(disp_hardpoint_axial)
+
+        # Note the negative piston means the +z
+        z = -dof_axial[2]
+
+        # Calculate the (rx, ry) in radian
+        mat_axial_3_points = np.array([[0, 0, 1], [1, 0, 1], [0, 1, 1]])
+        z_axial_3_points = mat_axial_3_points.dot(dof_axial)
+
+        mat_axial_3_points_updated = mat_axial_3_points.copy().astype(float)
+        mat_axial_3_points_updated[:, 2] = z_axial_3_points
+
+        r1c = mat_axial_3_points_updated[1, :] - mat_axial_3_points_updated[0, :]
+        r2 = mat_axial_3_points_updated[2, :] - mat_axial_3_points_updated[0, :]
+        cross_r1c_r2 = np.cross(r1c, r2)
+        cross_r1c_r2_normalized = cross_r1c_r2 / np.linalg.norm(cross_r1c_r2)
+
+        vector_rxryrz = np.cross(np.array([0, 0, 1]), cross_r1c_r2_normalized)
+        norm_vector_rxryrz = np.linalg.norm(vector_rxryrz)
+        if norm_vector_rxryrz == 0:
```

Comment:
Will this ever be `True`? The variable `norm_vector_rxryrz` comes from `norm_vector_rxryrz = np.linalg.norm(vector_rxryrz)`, which I think is operating on floats. I wonder if there's any chance that it will actually return 0.

Reply:
I think this is because of the following equation:

```
vector_rxryrz = np.cross(np.array([0, 0, 1]), cross_r1c_r2_normalized)
```

The original developer wanted to consider the possibility that the `cross_r1c_r2_normalized` might be a zero vector to make the consideration complete.